lookout/style/format/rules.py

```diff
@@ -390,11 +404,69 @@ def _compile_rules(cls, rules: Sequence[Rule]) -> CompiledRulesType:
                 tuple(cls.CompiledNegatedRules(*v) for v in vr))
         return compiled_attrs
 
+    @staticmethod
+    def _get_parents(tree: DecisionTreeClassifier) -> Dict[int, int]:
+            parents = {x: i for i, x in enumerate(tree.children_left) if x != Tree.TREE_LEAF}
+            parents.update(
+                {x: i for i, x in enumerate(tree.children_right) if x != Tree.TREE_LEAF})
+            return parents
+
     @classmethod
-    def _prune_branches(cls, rules: Sequence[Rule],
-                        X: numpy.ndarray, Y: numpy.ndarray) -> List[Rule]:
-        # TODO(vmarkovtsev): implement this function
-        return rules
+    def reduced_error_prune(cls, model: DecisionTreeClassifier, validation_X, validation_y,
+                            possible_step_score_drop=0, possible_max_score_drop=0, verbose=False):
+        def _prune_tree(tree, node_to_prune):
+            backup = (tree.children_left[node_to_prune],
+                      tree.children_right[node_to_prune],
+                      tree.feature[node_to_prune])
+            tree.children_left[node_to_prune] = Tree.TREE_LEAF
+            tree.children_right[node_to_prune] = Tree.TREE_LEAF
+            tree.feature[node_to_prune] = Tree.TREE_UNDEFINED
+            return backup
+
+        def _restore_tree(tree, node, backup):
+            tree.children_left[node], tree.children_right[node], tree.feature[node] = backup
+
+        model = deepcopy(model)
+        tree = model.tree_
+        init_score = curent_score = model.score(validation_X, validation_y)
+        changes = True
+        bad_change = set()
+        parents = cls._get_parents(tree)
+        leafs = list(numpy.where(tree.children_left == Tree.TREE_LEAF)[0])
```

Comment:
`leafs` should be `leaves`

Reply:
I need spell checker from @irinakhismatullina as soon as possible :sweat_smile:

Reply:
cc @eiso